### PR TITLE
Minion subnet

### PR DESCRIPTION
**Pull request by mmussomele**
This patch changes IP assignment so that each minion gets it's own subnet and assigns IPs locally.

Unit tested and works on sparkPI.spec

### CODE REVIEW

File: minion/etcd/network.go [hunk 1]
```diff
@@ -125,6 +142,50 @@ func runNetwork(conn db.Conn, store Store) {
...
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func generateSubnet(store Store) net.IP {
+	// XXX: Replace this with a less random selection process. i.e. select a random
+	// element from a set of untaken subnets
+	for {
+		subnet := randomMinionSubnet()
+		err := store.Create(subnetStore+strconv.Itoa(int(subnet)), "", 0)
```
Comment (kklin) [inline minion/etcd/network.go]: Could we have `randomMinionSubnet` return a `net.IP`, and just store the String representation in etcd?

File: minion/etcd/network.go [hunk 2]
```diff
@@ -7,6 +7,7 @@ import (
 	"math/rand"
```
Comment (ejj) [inline minion/etcd/network.go]: Commit title reads a bit weird to me.  Also I think it's too long.  Perhaps just "Make workers assign container IPs"

File: minion/etcd/network.go [hunk 3]
```diff
@@ -22,11 +23,25 @@ const (
 	minionDir      = "/minion"
 	labelToIPStore = minionDir + "/labelIP"
 	containerStore = minionDir + "/container"
+	subnetStore    = minionDir + "/subnets/"
+
+	privateMaskBits  = 8
+	privateSubnetInt = 0x0a000000
```
Comment (ejj) [inline minion/etcd/network.go]: I don't much like the names of these variables.

I think it might be clearer if you had `quiltPrefix`, `quiltMask`, `subPrefix`, `subMask` and possibly `labelPrefix`.

Also I'd use the `net` utilities to express these rather than writing out the hex.  Also I think I'd story the prefix as a mask instead of as `20` or whatever.

File: minion/etcd/network.go [hunk 4]
```diff
@@ -76,6 +91,7 @@ func wakeChan(conn db.Conn, store Store) chan struct{} {
 }
 
 func runNetwork(conn db.Conn, store Store) {
+	subnet := generateSubnet(store)
```
Comment (ejj) [inline minion/etcd/network.go]: Should the masters be generating a subnet for themselves?

File: minion/etcd/network.go (hunk 2 above)
Comment (ejj) [inline minion/etcd/network.go]: +1

Comment (ejj) [inline minion/etcd/network.go]: I also think that doing a simple Create isn't sufficient here.  We really need this subnet key to disappear from the etcd in the event the minion dies (so we don't end up leaking subnets).

**Comment by mmussomele** [comment]
Changes made and tested.

File: minion/etcd/etcd.go [hunk 5]
```diff
@@ -20,6 +20,7 @@ type Store interface {
 	Create(path, value string, ttl time.Duration) error
 	Update(path, value string, ttl time.Duration) error
 	Set(path, value string, ttl time.Duration) error
+	Refresh(path, value string, ttl time.Duration) error
```
Comment (ejj) [inline minion/etcd/etcd.go]: Eh . . . I think you actually uncovered a rather serious bug in the etcd elector code. `minion/etcd/elector` uses the Update() function to refresh the leader election key which is technically wrong.  If, for whatever reason, some other node has taken control over the key, we shouldn't overwrite it, but instead fail to update as your Refresh function here does.

Since that's such a serious bug, I think we should call it out with it's own commit.  Specifically, I think the addition of the Refresh function, and the associated fix to minion/etcd/elector.go should be in it's own commit before the rest of the series.  Should be easy, just moving code around.

Comment (ejj) [inline minion/etcd/etcd.go]: Bonus points if you can write a unit test that catches the problem btw =)

File: minion/etcd/minion.go [hunk 6]
```diff
@@ -11,12 +15,30 @@ import (
 	log "github.com/Sirupsen/logrus"
 )
 
-const timeout = 30
+const (
+	minionTimeout = 30
+	subnetStore   = minionDir + "/subnets"
+	subnetTTL     = 30 * time.Minute
```
Comment (ejj) [inline minion/etcd/minion.go]: That's a bit excessive . . .  Maybe 5 minutes?

File: minion/etcd/minion.go [hunk 7]
```diff
@@ -11,12 +15,30 @@ import (
...
+)
+
+var (
+	// Store in a variable so we can change it for unit tests
+	subnetAttempts = 1000
+
+	quiltPrefix = net.IPv4(10, 0, 0, 0)
```
Comment (ejj) [inline minion/etcd/minion.go]: Minor nit, but it'd be nice to have a comment above this saying "10.0.0.0/8".  The CIDRMask function is a bit odd looking so having the comment would clarify a lot.

File: minion/etcd/minion.go [hunk 8]
```diff
@@ -26,7 +48,7 @@ func runMinionSync(conn db.Conn, store Store) {
 func readMinion(conn db.Conn, store Store) {
 	tree, err := store.GetTree("/minion/nodes")
 	if err != nil {
-		log.WithError(err).Warning("Failed to get minions form Etcd.")
+		log.WithError(err).Warning("Failed to get minions from Etcd.")
```
Comment (ejj) [inline minion/etcd/minion.go]: Separate patch.

File: minion/etcd/minion.go [hunk 9]
```diff
@@ -119,7 +141,90 @@ func writeMinion(conn db.Conn, store Store) {
...
 
 	key := "/minion/nodes/" + minion.PrivateIP
-	if err := store.Set(key, string(js), timeout*time.Second); err != nil {
+	if err := store.Set(key, string(js), minionTimeout*time.Second); err != nil {
 		log.Warning("Failed to update minion node in Etcd: %s", err)
 	}
 }
+
+func syncSubnet(conn db.Conn, store Store) {
+	minion, err := conn.MinionSelf()
```
Comment (ejj) [inline minion/etcd/minion.go]: Random unrelated side note.  I wonder if MinionSelf() should panic if there is no self.  Would save a lot of useless error checking for something that can only occur if we have a bug anyways.

File: minion/etcd/minion.go [hunk 10]
```diff
@@ -119,7 +141,90 @@ func writeMinion(conn db.Conn, store Store) {
...
+		time.Sleep(time.Minute)
+	}
+
+	conn.Transact(func(view db.Database) error {
+		view.Commit(minion)
+		return nil
+	})
+}
+
+func generateSubnet(store Store, minion db.Minion) (string, error) {
```
Comment (ejj) [inline minion/etcd/minion.go]: Minor nit, but the file would read a bit more easily if this function came before the sync function.  Logically this code makes more sense to read first.

File: minion/etcd/minion.go [hunk 11]
```diff
@@ -119,7 +141,90 @@ func writeMinion(conn db.Conn, store Store) {
...
+	// XXX: Replace this with a less random selection process. i.e. select a random
```
Comment (ejj) [inline minion/etcd/minion.go]: Meh.  I think it's fine.

Though here's an idea for an algorithm that's simple if you really care.
```
cur = randomSubnet()
For subnetAttempts:
    check(cur)
    if failed:
         cur = (cur + 1) % range
```

Comment (ejj) [inline minion/etcd/minion.go]: To be clear, I think you should either remove the comment or implement the new algorithm.  I don't care which.

File: minion/etcd/minion.go [hunk 12]
```diff
@@ -119,7 +141,90 @@ func writeMinion(conn db.Conn, store Store) {
...
+		subnet := randomMinionSubnet()
+		err := store.Create(subnetKey(subnet), minion.PrivateIP, subnetTTL)
+		if err == nil {
+			return subnet, nil
+		}
+		log.WithError(err).WithField("subnet",
+			subnet).Debug("Subnet taken, trying again.")
+	}
+
+	return "", errors.New("failed to create subnet")
```
Comment (ejj) [inline minion/etcd/minion.go]: I like the word `allocate` instead of `create` here.

File: minion/etcd/minion.go [hunk 13]
```diff
@@ -119,7 +141,90 @@ func writeMinion(conn db.Conn, store Store) {
...
+}
+
+func randomMinionSubnet() string {
+	minionBits, _ := subMask.Size()
+	quiltBits, _ := quiltMask.Size()
+	subnetBits := minionBits - quiltBits
+	maxIP := uint32(0x7fffffff >> uint(31-subnetBits))
```
Comment (ejj) [inline minion/etcd/minion.go]: I think all of this would be simpler if you just choose a random 32bi `x` and then did `randomSubnet = x & (~mask) | prefix`

File: minion/etcd/minion_test.go [hunk 14]
```diff
@@ -160,6 +161,35 @@ func TestFilter(t *testing.T) {
 	}
 }
 
+// Test that GenerateSubnet generates valid subnets, and passively test that it can
+// generate enough unique subnets by generating 3500 out of the possible 4095
+func TestGenerateSubnet(t *testing.T) {
```
Comment (ejj) [inline minion/etcd/minion_test.go]: Test for sync subnet?

Comment (ejj) [inline minion/etcd/minion_test.go]: Basically all the new code in this file should have coverage IMO.

File: minion/etcd/network.go [hunk 15]
```diff
@@ -24,9 +24,13 @@ const (
...
 
-// We store rand.Uint32() in a variable so it's easily mocked out by the unit tests.
-// Nondeterminism is hard to test.
-var rand32 = rand.Uint32
+var (
+	// We store rand.Uint32() in a variable so it's easily mocked out by the unit
+	// tests. Nondeterminism is hard to test.
+	rand32 = rand.Uint32
+
+	labelPrefix = net.IPv4(10, 0, 0, 0) // Labels get their own /20
```
Comment (ejj) [inline minion/etcd/network.go]: Can we put all of the prefixes in some central constants package?  I'm worried about it getting out of sync.

File: minion/etcd/network.go [hunk 16]
```diff
@@ -173,48 +199,44 @@ func updateEtcdContainer(s Store, etcdData storeData, containers []db.Container)
...
-	score := func(left, right interface{}) int {
-		return containerJoinScore(left.(storeContainer), right.(storeContainer))
-	}
-	pairs, _, _ := join.Join(dbContainers, etcdData.containers, score)
+	dbContainerSlice := dbSliceToStoreSlice(containers)
 
-	newIPMap := map[string]string{}
-	for _, c := range dbContainers {
-		newIPMap[string(c.StitchID)] = ""
+	// Even if we aren't assigning IPs, we still need to sync up with the IPs
```
Comment (ejj) [inline minion/etcd/network.go]: Meh, I don't like this comment.  A random reader of the code in the future won't get the "Even if we aren't assigning IPs reference".

File: minion/etcd/network.go [hunk 17]
```diff
@@ -332,6 +356,66 @@ func updateWorkerDBC(view db.Database, self db.Minion, etcdData storeData) {
 
 		view.Commit(dbc)
 	}
+
+	if self.Subnet == "" {
+		return etcdData, nil
+	}
+
+	oldContainers := view.SelectFromContainer(nil)
```
Comment (ejj) [inline minion/etcd/network.go]: I don't really like that each worker is attempting to update the etcd container's file.  (1) it leads to this rather awkward merging code which is somewhat error-prone and (2) having all the nodes hit this one file is going to cause either failed writes, or workers over-writing each others updates.

I think it would be much better if we had a directory which contained a key per worker that had a json map from stitch ID to Ip for those containers running on that worker.  Then each worker, when populating it's local database, would read all the keys in that directory and do a join to assign the Ips locally.

File: minion/etcd/network.go [hunk 18]
```diff
@@ -411,63 +495,70 @@ func containerJoinScore(left, right storeContainer) int {
 // syncIPs takes a map of IDs to IPs and creates an IP address for every entry that's
 // missing one.
 func syncIPs(ipMap map[string]string, prefixIP net.IP) {
```
Comment (ejj) [inline minion/etcd/network.go]: Is there any way we could somehow pull this Ip management code out into a library that is used by bot the containers and labels?

File: minion/constants/constants.go [hunk 19]
```diff
@@ -0,0 +1,24 @@
+package constants
```
Comment (ejj) [inline minion/constants/constants.go]: Should this be in the previous, or it's own patch?

File: stitch/stitch.go [hunk 20]
```diff
@@ -91,6 +93,11 @@ func toStitch(parsed []ast) (Stitch, error) {
 		return Stitch{}, err
 	}
 
+	if tooManyWorkers(*ctx.machines) {
```
Comment (ejj) [inline stitch/stitch.go]: @kklin, what do you think of this?  I'm not sure this should be in stitch as it's very QRI specific.  I might be inclined to drop the patch entirely.  Specifying workers is in some ways like specifying too many containers, we'll make a best effort, but may fail.

Comment (ejj) [inline stitch/stitch.go]: Sure that feels much better to me.

File: minion/etcd/minion.go (hunk 9 above)
Comment (mmussomele) [inline minion/etcd/minion.go]: Minions won't have a self until they receive their initial config, so they'll be running first.

File: minion/constants/constants.go (hunk 19 above)
Comment (mmussomele) [inline minion/constants/constants.go]: There's no reason for them to be exported until this patch

File: minion/etcd/network.go (hunk 15 above)
Comment (mmussomele) [inline minion/etcd/network.go]: It seems to make sense to put them with the IP management library for the minion.

**Comment by mmussomele** [comment]
Changed made and code tested. Confirmed to work on sparkPI.

**Comment by mmussomele** [comment]
Note: There's a performance issue with adding TTL support to etcd_mock, I'm working on diagnosing it.

File: minion/etcd/minion.go (hunk 13 above)
Comment (mmussomele) [inline minion/etcd/minion.go]: Didn't quite do this, but I found a way that I think is simpler

**Comment by mmussomele** [comment]
Ready for review.

File: minion/etcd/network.go [hunk 21]
```diff
@@ -166,48 +185,55 @@ func updateEtcdContainer(s Store, etcdData storeData, containers []db.Container)
...
 
-	ip.Sync(newIPMap, net.IPv4(10, 0, 0, 0))
-	for i := range dbContainers {
-		dbContainers[i].IP = newIPMap[string(dbContainers[i].StitchID)]
+	dbContainerSlice := dbSliceToStoreSlice(containers)
+	for i := range dbContainerSlice {
+		id := dbContainerSlice[i].StitchID
+		dbContainerSlice[i].IP = ipMap[strconv.Itoa(id)]
 	}
+	sort.Sort(storeContainerSlice(dbContainerSlice))
```
Comment (ejj) [inline minion/etcd/network.go]: I'd consider putting the sort in the writeDiff function.  That way it's a little bit more clear when you read the writeDiff why the string comparison is supposed to work

File: minion/etcd/network.go [hunk 22]
```diff
@@ -325,6 +353,78 @@ func updateWorkerDBC(view db.Database, self db.Minion, etcdData storeData) {
 
 		view.Commit(dbc)
 	}
+
```
Comment (ejj) [inline minion/etcd/network.go]: I'm sure this code works, but I think it's a bit difficult to reason about the fact that `etcdc` has an IP field given that the worker should be completely ignoring it in this code.  I think in this patch, we should remove the IP from etcdc just for clarity -- plus it removes any possibility of etcdc and the dbc from disagreeing on the IP for a particular container, and that causing confusion.

File: minion/etcd/network.go [hunk 23]
```diff
@@ -325,6 +353,78 @@ func updateWorkerDBC(view db.Database, self db.Minion, etcdData storeData) {
...
+	subnet := ip.Parse(self.Subnet, ip.QuiltPrefix, ip.QuiltMask)
+
+	oldContainers := view.SelectFromContainer(nil)
+	newIPMap, err := updateContainerIP(etcdData, oldContainers, subnet, self, store)
+	if err != nil {
+		return
+	}
+
+	for _, c := range oldContainers {
+		if ipAddr, ok := newIPMap[strconv.Itoa(c.StitchID)]; ok && ipAddr != "" {
```
Comment (ejj) [inline minion/etcd/network.go]: Why would this ever happen?  Isn't the source of truth on the IP the database code? I.E. shouldn't a particular container's IP be fixed by the code that's doing the assignment?

File: minion/etcd/network.go [hunk 24]
```diff
@@ -383,6 +483,19 @@ func updateDBLabels(view db.Database, etcdData storeData) {
...
+	}
+
+	for k, v := range first {
+		if val, ok := second[k]; !ok || v != val {
+			return false
+		}
+	}
+	return true
+}
+
```
Comment (ejj) [inline minion/etcd/network.go]: Btw, I think containerJoinScore should change (possibly in an earlier patch).

Basically, I think we should add a patch before this one that says a given containers Ip address should never change. Thus the join score shouldn't take into account the IP address, (or rather the score should be -1 if the IP is different).

File: minion/etcd/network.go (hunk 23 above)
Comment (ejj) [inline minion/etcd/network.go]: Ah, I guess it would happen because we're assigning IPs in updateContainer IP?  So this code would go away later?  I think I'd add a comment to that affect at least.

File: minion/etcd/network.go [hunk 25]
```diff
@@ -325,6 +353,78 @@ func updateWorkerDBC(view db.Database, self db.Minion, etcdData storeData) {
...
+		return oldIPMap, nil
+	}
+
+	jsonData, err := json.Marshal(newIPMap)
+	if err != nil {
+		log.WithError(err).Error("Failed to marshal minion container IP map")
+		return oldIPMap, err
+	}
+
+	err = store.Set(path.Join(minionIPStore, self.PrivateIP), string(jsonData), 0)
```
Comment (ejj) [inline minion/etcd/network.go]: What if the minion dies . . . do we leak this key or does it get cleaned up later?

I wonder if the minion key in minion/etcd/minion.go should be a directory instead?

Sigh, an etcd3 lease would simplify a lot of this.  Too bad we didn't get the update done.

File: minion/etcd/network.go [hunk 26]
```diff
@@ -383,6 +483,19 @@ func updateDBLabels(view db.Database, etcdData storeData) {
 	}
 }
 
+func strMapEquals(first, second map[string]string) bool {
```
Comment (ejj) [inline minion/etcd/network.go]: util has an implementation of this.

File: minion/ip/ip.go [hunk 27]
```diff
@@ -15,56 +15,93 @@ var (
...
+func FromInt(ip32 uint32) net.IP {
+	b := make([]byte, 4)
+	binary.BigEndian.PutUint32(b, ip32)
+	return net.IP(b)
+}
+
+// Random selects a random IP address in the given prefix and mask that isn't already
+// present in conflicts.
+// Returns 0 on failure.
+func Random(conflicts map[string]struct{}, pre net.IP, subnetMask net.IPMask) net.IP {
```
Comment (ejj) [inline minion/ip/ip.go]: Probably belongs in a separate patch, but I think since we're working with a smaller range here, we should ditch this Random allocation logic.  My preference would be to search the entire subnet for an IP linearly starting from a random location.

File: minion/etcd/network.go (hunk 21 above)
Comment (mmussomele) [inline minion/etcd/network.go]: Actually this is the only place it's used, I think I'm gonna ditch it

Comment (ejj) [inline minion/etcd/network.go]: Don't you need the sort for the diff to work?

File: minion/etcd/network.go (hunk 22 above)
Comment (mmussomele) [inline minion/etcd/network.go]: So storeContainers still need the IP field since it makes the master logic simpler. I'll just have the worker ignore it and overwrite it if the ipMap updateContainerIPs returns has a conflicting value.

File: minion/etcd/network.go (hunk 23 above)
Comment (mmussomele) [inline minion/etcd/network.go]: so this code changed slightly because of the above comment, I think this is clear:
```
for _, c := range oldContainers {
                 ipAddr, ok := newIPMap[strconv.Itoa(c.StitchID)]
                 if ok && ipAddr != c.IP {
                         c.IP = ipAddr
                         c.Mac = ip.ToMac(ipAddr)
                         view.Commit(c)
                 }
         }
```

File: minion/ip/ip.go (hunk 27 above)
Comment (mmussomele) [inline minion/ip/ip.go]: hmmm, problem is the Master updates the storeContainers and they'll all have "" for an IP until it syncs up, so all the container will just be destroyed if we do that (none will join up)

File: minion/etcd/network.go (hunk 24 above)
Comment (mmussomele) [inline minion/etcd/network.go]: hmmm, problem is the Master updates the storeContainers and they'll all have "" for an IP until it syncs up, so all the container will just be destroyed if we do that (none will join up)

Comment (mmussomele) [inline minion/etcd/network.go]: ah you know what, I'll only return -1 if they disagree and the etcd container actually has an IP address

**Comment by mmussomele** [comment]
Changes made, unit tested, and sparkPI tested.

**Comment by mmussomele** [comment]
@ejj ready for another review

File: minion/etcd/network.go (hunk 22 above)
Comment (ejj) [inline minion/etcd/network.go]: Yeah I don't quite understand this point still.  Perhaps we should talk about it on slack or in person.  If it really does make the master simpler I think it's fine, but I'd like to understand it a bit better.

File: minion/etcd/minion.go [hunk 28]
```diff
@@ -118,8 +155,106 @@ func writeMinion(conn db.Conn, store Store) {
...
+	err := store.Refresh(subnetKey(minion.Subnet), minion.PrivateIP,
+		subnetTTL)
+	if err == nil {
+		return minion
+	}
+	log.WithError(err).Infof("Failed to refresh subnet '%s', "+
+		"generating a new one.", minion.Subnet)
+
+	// Invalidate the subnet until we get a new one.
+	conn.Transact(func(view db.Database) error {
```
Comment (ejj) [inline minion/etcd/minion.go]: This logic repeats just below.  I think you could pull it out into some sort of setMinionSubnet function.

File: minion/etcd/minion.go [hunk 29]
```diff
@@ -118,8 +155,106 @@ func writeMinion(conn db.Conn, store Store) {
...
+	})
+
+	// If we failed to refresh, someone took our subnet or we never had one.
+	for {
+		minion.Subnet, err = generateSubnet(store, minion)
+		if err == nil {
+			break
+		}
+
+		log.WithError(err).Error("Failed to allocate subnet, " +
```
Comment (ejj) [inline minion/etcd/minion.go]: Warn.

File: minion/etcd/minion.go [hunk 30]
```diff
@@ -2,29 +2,65 @@ package etcd
...
+	minionTimeout = 30
+	subnetStore   = minionDir + "/subnets"
+	selfNode      = "self"
+)
+
+var (
+	// Store in variables so we can change them for unit tests
+	subnetAttempts = 1000
+	syncWait       = 15 * time.Second
+	ttlWait        = time.Minute
```
Comment (ejj) [inline minion/etcd/minion.go]: I'm not sure we really need ttlWait at all.

File: minion/etcd/minion.go [hunk 31]
```diff
@@ -118,8 +155,106 @@ func writeMinion(conn db.Conn, store Store) {
...
+	// If we failed to refresh, someone took our subnet or we never had one.
+	for {
+		minion.Subnet, err = generateSubnet(store, minion)
+		if err == nil {
+			break
+		}
+
+		log.WithError(err).Error("Failed to allocate subnet, " +
+			"trying again in a minute")
+		sleep(ttlWait)
```
Comment (ejj) [inline minion/etcd/minion.go]: Not clear why we need to sleep this long.  If it failed to to a race condition, I think we should just wait a second or two and be fine.

File: minion/etcd/minion.go [hunk 32]
```diff
@@ -2,29 +2,65 @@ package etcd
...
+const (
+	minionTimeout = 30
+	subnetStore   = minionDir + "/subnets"
+	selfNode      = "self"
+)
+
+var (
+	// Store in variables so we can change them for unit tests
+	subnetAttempts = 1000
+	syncWait       = 15 * time.Second
```
Comment (ejj) [inline minion/etcd/minion.go]: I also generally like these wait timers to be expressed as a fraction of the TTL.  I.e. synWait should be subnetTTL / 4 or something.
